airflow/jobs.py

```diff
@@ -1530,9 +1530,25 @@ def process_file(self, file_path, pickle_dags=False, session=None):
             dag = dagbag.dags[ti_key[0]]
             task = dag.get_task(ti_key[1])
             ti = models.TaskInstance(task, ti_key[2])
-            # Task starts out in the scheduled state. All tasks in the
-            # scheduled state will be sent to the executor
-            ti.state = State.SCHEDULED
+
+            ti.refresh_from_db(session=session, lock_for_update=True)
+            # We can defer checking the task dependency checks to the worker themselves
+            # since they can be expensive to run in the scheduler.
+            dep_context = DepContext(deps=QUEUE_DEPS, ignore_task_deps=True)
+
+            # Only schedule tasks that have their dependencies met, e.g. to avoid
+            # a task that recently got it's state changed to RUNNING from somewhere
+            # other than the scheduler from getting it's state overwritten.
+            # TODO(aoen): It's not great that we have to check all the task instance
+            # dependencies twice; once to get the task scheduled, and again to actually
+            # run the task. We should try to come up with a way to only check them once.
```

Comment:
As things are today, I think it's a necessity since the conditions may change between the time the job is scheduled and started (which typically happens on different machines/interpreters). Maybe there is a [new] special context like distributed scheduling where a finishing task could play the role of the scheduler and schedule one of its dependents job while re-using the worker slot, but the simpler the better. For consistency and simplicity: measure twice, cut once.